### DISCUSSION TITLE
Evaluate Test2::Tools::Spec

### DISCUSSION THREAD

**Discussion by bschmalhofer** [RotherOSS/otobo#2185] [Ideas]
[Test2::Tools::Spec](https://metacpan.org/release/EXODIST/Test2-Suite-0.000145/view/lib/Test2/Tools/Spec.pm) looks like an useful way of writing tests. But it is not obvious whether it is a good fit for OTOBO. Maybe rewritiing a large test with Test2::Tools::Spec will provide some insight.

**Comment by mario-minati** [comment]
I would suggest to setup the new pipelines first and checkout what further benefits can be found in rewriting tests.

**Comment by bschmalhofer** [comment]
Test2::Tools::Spec is solely concerned about how the test is written. The integration in workflows does not really care, as the output from the script is always TAP, or Test2-Events.

**Comment by stefanhaerter** [comment]
Personally, I would appreciate some kind of best practice or structured way to write tests, since I sometimes feel a bit lost within this topic. I had a quick look at the module page: Can you tell whether this is a framework for grouping existing tests or for writing tests?

**Reply by bschmalhofer** [reply]
Test2::Tools::Spec is mostly for grouping tests and for neater structuring the setup and teardown of the tests. E.g. like 

1. create a test ticket
2. do a bunch of tests on the ticket
3. close or remove the ticket

The assertions are not different at all. That is, the test Functions from Test2::V0 und Kernel::System::UnitTest::Driver are still used.

I propose to restructute a complicated test script and then see whether the structure of the script becomes more maintainable when using Test2::Tools::Spec.

**Reply by stefanhaerter** [reply]
The concept sounds interesting to me. I would second to rewriting an existing test and evaluate the result.
